pkg/manager/manager.go

```diff
@@ -0,0 +1,187 @@
+package manager
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	dhclient "github.com/digineo/go-dhclient"
+	"github.com/kamhlos/upnp"
+	"github.com/plunder-app/kube-vip/pkg/bgp"
+	"github.com/plunder-app/kube-vip/pkg/cluster"
+	"github.com/plunder-app/kube-vip/pkg/kubevip"
+	log "github.com/sirupsen/logrus"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const plunderLock = "plndr-svcs-lock"
+
+var signalChan chan os.Signal
+
+// OutSideCluster allows the controller to be started using a local kubeConfig for testing
+var OutSideCluster bool
+
+type kubernetesServices struct {
+	Services []service `json:"services"`
+}
+
+// Manager degines the manager of the load-balancing services
+type Manager struct {
+	clientSet *kubernetes.Clientset
+	configMap string
+	config    *kubevip.Config
+
+	// Manager services
+	service bool
+
+	// Keeps track of all running instances
+	serviceInstances []Instance
+
+	// Additional functionality
+	upnp *upnp.Upnp
+	//BGP Manager, this is a singleton that manages all BGP advertisements
+	bgpServer *bgp.Server
+}
+
+type dhcpService struct {
+	// dhcpClient (used DHCP for the vip)
+	dhcpClient    *dhclient.Client
+	dhcpInterface string
+}
+
+// Instance defines an instance of everything needed to manage a vip
+type Instance struct {
+	// Virtual IP / Load Balancer configuration
+	vipConfig kubevip.Config
+	// Kubernetes service mapping
+	service service
+	// cluster instance
+	cluster cluster.Cluster
+
+	// Custom settings
+	dhcp *dhcpService
+}
+
+// TODO - call from a package (duplicated struct in the cloud-provider code)
+type service struct {
+	Vip  string `json:"vip"`
+	Port int    `json:"port"`
+	UID  string `json:"uid"`
+	Type string `json:"type"`
+
+	ServiceName string `json:"serviceName"`
+}
+
+// SetControlPane determines if the control plane should be enabled
+// func (sm *Manager) SetControlPane(enable bool) {
+// 	sm.controlPane = enable
+// }
+
+// New will create a new managing object
+func New(configMap string, config *kubevip.Config) (*Manager, error) {
+	var clientset *kubernetes.Clientset
+	if OutSideCluster == false || !config.EnableControlPane {
+		// This will attempt to load the configuration when running within a POD
+		cfg, err := rest.InClusterConfig()
+		if err != nil {
+			return nil, fmt.Errorf("error creating kubernetes client config: %s", err.Error())
+		}
+		clientset, err = kubernetes.NewForConfig(cfg)
+
+		if err != nil {
+			log.Debugln("Using incluster Kubernetes configuration")
+			return nil, fmt.Errorf("error creating kubernetes client: %s", err.Error())
+		}
+		// use the current context in kubeconfig
+	} else {
+		// Check for file existing
+		// First for default path on control plane
+		// /etc/kubernetes/admin.conf
+		var configPath string
+		var cfg *rest.Config
+		var err error
+
+		configPath = "/etc/kubernetes/admin.conf"
```

Comment:
This is unfortunately needed for static pods for the control plane. They have no token and operate in the same way as things like the api-server.

Reply:
Yeah, I think it's fine for a short term solution, but longer term it might be good to think about something like:
- Create a net new kubeconfig using a client certificate that is scope limited compared to the admin kubeconfig (similar to what is done for the controller-manager, scheduler, kubelet, etc)
- Figure out how to refresh and update above certificate/kubeconfig

It should be relatively easy with Cluster API, as we have access to all the primitives we need to make this happen, but need to figure it out for non-cluster api use cases.

We could potentially attempt to replicate kubelet client certificate rotation: https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet-tls-bootstrapping/#certificate-rotation. I suspect it will require some additional RBAC injected into the cluster to make auto approval of the CSR requests work, similar to what is done in kubeadm: https://github.com/kubernetes/kubernetes/blob/c1f36fa6f28d3618c03b65799bc3f58007624e5f/cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go